File: 07c_Recycler_View/app/src/main/java/com/example/a7c_recycler_view/MainActivity.java
package com.example.a7c_recycler_view;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;

public class MainActivity extends AppCompatActivity {
    RecyclerView recyclerView;
    String[] items = {"Item 1", "Item 2", "Item 3", "Item 1", "Item 2", "Item 3", "Item 1", "Item 2", "Item 3","Item 1", "Item 2", "Item 3", "Item 1", "Item 2", "Item 3", "Item 1", "Item 2", "Item 3"};
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        recyclerView =findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        CustomAdapter customAdapter = new CustomAdapter(items);
        recyclerView.setAdapter(customAdapter);
    }
}

File: 32a_Retrofit_Demo/app/src/main/java/com/example/a32a_retrofit_demo/MainActivity.kt
package com.example.a32a_retrofit_demo

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {
    private lateinit var rvNews : RecyclerView
    private val articles = mutableListOf<Article>()
    lateinit var loader : ProgressBar
//    private var articles =
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        loader = findViewById(R.id.loader)
        rvNews = findViewById(R.id.rvNews)
        loader.visibility = View.VISIBLE
        rvNews.adapter = NewsAdapter( articles,this@MainActivity)
        rvNews.layoutManager = LinearLayoutManager(this@MainActivity)

        fetchNews()
    }

    fun fetchNews(){
        val news = NewsService.INSTANCE.getNews()
        news.enqueue(object : Callback<News>{
            override fun onResponse(call: Call<News>, response: Response<News>) {
                Log.d("MyTag", response.body().toString())
                loader.visibility = View.INVISIBLE
                if(response.body()!=null){
                    articles.addAll(response.body()!!.articles)
                    rvNews.adapter?.notifyDataSetChanged()
//                    rvNews.adapter = NewsAdapter(response.body()!!.articles, this@MainActivity)
//                    rvNews.layoutManager = LinearLayoutManager(this@MainActivity)
                }
            }

            override fun onFailure(call: Call<News>, t: Throwable) {
                Log.d("MyTag", "${t.message}")
                loader.visibility = View.INVISIBLE
                Toast.makeText(this@MainActivity,"${t.message}", Toast.LENGTH_SHORT).show()
            }

        })
    }
}